controller/server_controller.py

```diff
@@ -125,6 +127,101 @@ def create_fn(labels, logger, name, namespace, spec, uid, **_):
     return {"createdResources": children_uids, "fullServerURL": get_urls(spec)[1]}
 
 
+@kopf.timer(
+    config.api_group,
+    config.api_version,
+    config.custom_resource_name,
+    interval=config.SESSION_IDLE_CHECK_INTERVAL_SECONDS,
+)
+def cull_idle_sessions(body, name, namespace, logger, **kwargs):
+    """
+    Check if a session is idle (has zero open connections in proxy and CPU is below
+    threshold). If the session is idle then annotate the jupyterserver resource with
+    the idle duration. If any sessions have been idle for long enough, then cull them.
+    """
+    try:
```

Comment:
In some cases (for example when having a long-running init container) - there is no main pod status. I'd check for this first and return right away, something like |
```
if not "mainPod" in body["status"]: 
   return
```